### ISSUE TITLE
Upsert giving 400 "member exists" error 

### ISSUE THREAD

**Issue by alexholmesv** [amro/gibbon#239]
Hi, 

When using the upsert method, I get a "member exists" 400 error from the API, even though that probably should only happen with create? At the moment I can't update any contacts in my list because I get this error.

**Error:**
`the server responded with status 400 @title="Member Exists", @detail="xxx@hotmail.com is already a list member. Use PUT to insert or update list members.", @body={"type"=>"http://developer.mailchimp.com/documentation/mailchimp/guides/error-glossary/", "title"=>"Member Exists", "status"=>400, "detail"=>"xxx@hotmail.com is already a list member. Use PUT to insert or update list members.", "instance"=>""}, @raw_body="{\"type\":\"http://developer.mailchimp.com/documentation/mailchimp/guides/error-glossary/\",\"title\":\"Member Exists\",\"status\":400,\"detail\":\"xxx@hotmail.com is already a list member. Use PUT to insert or update list members.\",\"instance\":\"\"}", @status_code=400`

**My method:**
def send_contact_to_mailing_list
      if account.mailchimp_api_key.nil? || get_subscriber_list_id.nil?
        puts "Not a mailchimp subscriber"
      else
        email = self.user.email
        gibbon = get_mailchimp_api_key
        md5_email = Digest::MD5.hexdigest(self.email)
        result = gibbon.lists(get_subscriber_list_id).members(md5_email.downcase).upsert(body: {
                                                                      email_address: self.email, 
                                                                      status: "subscribed", 
                                                                      merge_fields: {
                                                                        FNAME: self.name}
                                                                        })
        Rails.logger.info("Subscribed #{self.email} to MailChimp") if result
      end
  end

Is upsert not working anymore in the newest version of the gem? Or could this be a problem with the API?

**Comment by amro** [comment]
Upsert should work fine. I don't know if this is *the* issue, but I noticed you're MD5'ing the email then lower-casing it. That's not correct. The idea is to generate a hash / identifier that is normalized so plase try lowercasing the email then md5 hashing it instead of the other way around. Hopefully that helps.

**Comment by alexholmesv** [comment]
Thanks for the reply! 

It's definitely not the downcase issue, since I've tried it before the md5 hash (and even without), yet the error is still the same (but thanks for pointing it out anyway!). 

At a first glance, it would seem that the API is trying to create rather than upsert the contact, hence the "member exists" error. Not sure why this could be though...

**Comment by amro** [comment]
Okay -- one more thought. Try deleting the member from the list then using upsert(). IIRC this happens if a user has unsubscribed. In this case their status is unsubscribed but they're not deleted from the list and upsert() will return this. I could be mistaken.

https://developer.mailchimp.com/documentation/mailchimp/reference/lists/members/

**Comment by amro** [comment]
Did that help @alexholmesv ?

**Comment by alexholmesv** [comment]
Yep. I deleted the contact from Mailchimp and this time upsert worked fine to create the contact.

I did some more testing and realized this happens if I send the contact for a second time without editing the contact info. In this case, if I change the contact name to something else, upsert works fine. But if I send the contact again with upsert without editing the contact info, the error shows up. It seems the Mailchimp API won't let you upsert and replace with identical info.

In the meantime, since the contact is already there, I used a rescue block in my method to avoid getting this error.

Thanks for the support!
